tiledb/sm/query/readers/dense_reader.cc

```diff
@@ -1410,5 +1432,243 @@ Status DenseReader::add_extra_offset() {
   return Status::Ok();
 }
 
+template <class T>
+void DenseReader::fill_dense_coords(
+    const Subarray& subarray, const optional<std::vector<uint8_t>> qc_results) {
+  auto timer_se = stats_->start_timer("fill_dense_coords");
+
+  // Count the number of cells.
+  auto cell_num = subarray.cell_num();
+  if (qc_results.has_value()) {
+    auto func = [](uint64_t res, const uint8_t v) { return res + v; };
+    cell_num = std::accumulate(
+        qc_results.value().begin(), qc_results.value().end(), 0, func);
+  }
+
+  // Prepare buffers.
+  std::vector<unsigned> dim_idx;
+  std::vector<QueryBuffer*> buffers;
+  auto coords_it = buffers_.find(constants::coords);
+  auto dim_num = array_schema_.dim_num();
+  if (coords_it != buffers_.end()) {
+    // Check for overflow.
+    if (coords_it->second.original_buffer_size_ <
+        cell_num * array_schema_.cell_size(constants::coords)) {
+      read_state_.overflowed_ = true;
+      return;
+    }
+    buffers.emplace_back(&(coords_it->second));
+    dim_idx.emplace_back(dim_num);
+  } else {
+    for (unsigned d = 0; d < dim_num; ++d) {
+      const auto dim{array_schema_.dimension_ptr(d)};
+      auto it = buffers_.find(dim->name());
+      if (it != buffers_.end()) {
+        // Check for overflow.
+        if (it->second.original_buffer_size_ <
+            cell_num * array_schema_.cell_size(dim->name())) {
+          read_state_.overflowed_ = true;
+          return;
+        }
+        buffers.emplace_back(&(it->second));
+        dim_idx.emplace_back(d);
+      }
+    }
+  }
+
+  uint64_t qc_results_index = 0;
+  std::vector<uint64_t> offsets(buffers.size(), 0);
+  if (layout_ == Layout::GLOBAL_ORDER) {
+    fill_dense_coords_global<T>(
+        subarray, qc_results, qc_results_index, dim_idx, buffers, offsets);
+  } else {
+    assert(layout_ == Layout::ROW_MAJOR || layout_ == Layout::COL_MAJOR);
+    fill_dense_coords_row_col<T>(
+        subarray, qc_results, qc_results_index, dim_idx, buffers, offsets);
+  }
+
+  // Update buffer sizes.
+  for (size_t i = 0; i < buffers.size(); ++i) {
+    *(buffers[i]->buffer_size_) = offsets[i];
```

Comment:
This is pre-existing code, but want to make a specific note here that we need to move away from manually managing information at multiple levels like this. We're making a call to some functions which manipulate the buffer, and then updating the buffer sizes at the parent level. This should be handled by the operations we do on the buffer itself (or other replacement datatype).